src/scanner.js

```diff
@@ -0,0 +1,69 @@
+import taskcluster from 'taskcluster-client'
+import scopeUtils from 'taskcluster-lib-scopes'
+import User from './user'
+var debug = require('debug')('scanner');
+
+export default async function scanner(cfg, authorizers) {
+  // * get the set of identityProviderIds
+  // * for each:
+  //   * fetch all clients, sort by identity
+  //   * for each identity:
+  //     * get roles from providers, expand
+  //     * for each client in that identity:
+  //       * get, verify client.expandedScopes satisfied by identity's expandedScopes
+
+  // NOTE: this function performs once auth operation at a time.  It is better
+  // for scans to take longer than for the auth service to be overloaded.
+  let auth = new taskcluster.Auth({credentials: cfg.app.credentials});
+
+  // gather all identityProviderIds used in any authorizer
+  let identityProviders = new Set(
+      authorizers.map(authz => authz.identityProviders).reduce((a,b) => a.concat(b)))
+
+  // enumerate all clients for any of those identity providers
+  let clients = [];
+  for (let idp of identityProviders) {
+    clients = clients.concat(await auth.listClients({prefix: idp + "/"}));
+  }
+  clients.sort((a, b) => {
+    if (a.clientId < b.clientId) return -1;
+    if (a.clientId > b.clientId) return 1;
+    return 0;
```

Comment:
The sorting is so that we only look up each user once.  Since the identity is a prefix of the clientId, sorting is an easy way to ensure this.  I didn't know about `_.sortBy`, though -- that's great.